File: delnodeandreturnforest.cpp
// #include <vector>
// #include <unordered_set>
// using namespace std;

// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };

class Solution {
public:
    TreeNode* dfs(TreeNode* root, unordered_set<int>& to_delete, vector<TreeNode*>& forest) {
        if (root == nullptr) {
            return root;
        }
        root->left = dfs(root->left, to_delete, forest);
        root->right = dfs(root->right, to_delete, forest);

        if (to_delete.find(root->val) == to_delete.end()) {
            return root;
        }

        if (root->left != nullptr) {
            forest.push_back(root->left);
        }
        if (root->right != nullptr) {
            forest.push_back(root->right);
        }
        root->left = nullptr;
        root->right = nullptr;
        return nullptr;
    }

    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        unordered_set<int> set(to_delete.begin(), to_delete.end());
        vector<TreeNode*> forest;
        root = dfs(root, set, forest);
        if (root != nullptr) {
            forest.push_back(root);
        }
        return forest;
    }
};

File: delnodeandreturnforest.py
# from typing import List, Optional, Set

# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def dfs(self, root: Optional[TreeNode], to_delete: Set[int], forest: List[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return root
        root.left = self.dfs(root.left, to_delete, forest)
        root.right = self.dfs(root.right, to_delete, forest)

        if root.val not in to_delete:
            return root

        if root.left is not None:
            forest.append(root.left)
        if root.right is not None:
            forest.append(root.right)
        root.left = None
        root.right = None
        return None

    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        to_delete_set = set(to_delete)
        forest = []
        root = self.dfs(root, to_delete_set, forest)
        if root is not None:
            forest.append(root)
        return forest
